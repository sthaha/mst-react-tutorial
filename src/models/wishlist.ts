import {types, Instance} from 'mobx-state-tree';


export const WishlistItem = types
  .model({
    name: types.string,
    price: types.number,
    url: "",
  })
  .actions( self => {
    const changeName = (name: string) => {self.name = name}

    return {
      changeName,
      changePrice(x: number){
        self.price = x
      }
    }
  })


  type FilterFn = (x: IWishlistItem) => boolean


  export const Wishlist = types
    .model({
      items: types.optional(types.array(WishlistItem), [])
    })
    .actions(self => ({
      add(item: IWishlistItem){
        self.items.push(item)
      },
    }))
    .actions(self => ({
      fetch() {
        //setTimeout(() => { self.add({name: 'foobar', price: Math.random() * 100}) }, 1000)
      },
    }))
    .views(self => ({
      get totalPrice() {
        return self.items.reduce((sum, item) => sum + item.price, 0)
      },
      filter(fn: FilterFn) {
        return self.items.filter(x => fn(x))
      }
    }))


export interface IWishlistItem extends Instance<typeof WishlistItem> {}
export interface IWishlist extends Instance<typeof Wishlist> {}

