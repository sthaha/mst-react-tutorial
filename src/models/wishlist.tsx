import {types} from 'mobx-state-tree';

export const WishListItem = types
  .model({
    name: types.string,
    price: types.number,
    url: "",
  })
  .actions( self => {
    const changeName = (name: string) => {self.name = name}

    return {changeName}
  })




  export const WishList = types
    .model({
      items: types.optional(types.array(WishListItem), [])
    })
    .actions(self => ({
      add(item){
        self.items.push(item)
      },
    }))
    .views(self => ({
      get totalPrice(){
        return self.items.reduce((sum, x) => sum + x, 0)
      }
    }))

