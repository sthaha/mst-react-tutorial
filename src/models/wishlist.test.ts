import {getSnapshot, onSnapshot, onPatch} from 'mobx-state-tree';
import {WishList, WishListItem } from './wishlist';


it("can create an item", () => {
  const item = WishListItem.create({
    name: 'test',
    price: 12.5,
  })

  expect(item.name).toBe('test')
  expect(item.url).toBe('')
  expect(item.price).toBe(12.5)
})

it("can rename an items", () => {
  const item = WishListItem.create({
    name: 'test',
    price: 12.5,
  })

  expect(item.name).toBe('test')
  item.changeName("foobar");
  expect(item.name).toBe('foobar')
})


it ("can create empty wishlist", () => {
  const list = WishList.create({})
  expect(list.items.length).toBe(0)

})


it("can create a wishlist", () => {

  const item1 = WishListItem.create({
    name: 'test 1',
    price: 12.5,
  })

  const item2 = WishListItem.create({
    name: 'test 2',
    price: 12.5,
  })

  const list = WishList.create({
    items: [item1, item2]
  })
})


it ("can add items to wishlist", () => {
  const list = WishList.create({})
  list.add({
    name: "something",
    price: 1.56
  })

  const first = list.items[0]

  expect(first.name).toBe("something")
  expect(first.price).toBe(1.56)

  first.changeName("foobar")
  expect(first.name).toBe("foobar")


  expect(getSnapshot(list)).toMatchSnapshot()
})


it ("can store patches of changes to wishlist", () => {
  const list = WishList.create({})

  const states = []
  onSnapshot(list, s => states.push(s))

  const patches = []
  onPatch(list, p => patches.push(p))

  list.add({
    name: "something",
    price: 1.56
  })
  const first = list.items[0]

  expect(first.name).toBe("something")
  expect(first.price).toBe(1.56)

  expect(getSnapshot(list)).toMatchSnapshot()
  expect(patches).toMatchSnapshot()
})

it("can calculate total price", () => {
  const list = WishList.create();
  expect(list.totalPrice).toBe(0);
  list.add({
    name: 'foobar',
    price: 1.23
  })
  expect(list.totalPrice).toBe(1.23);

  list.add({
    name: 'another one',
    price: 2.36,
  })
  expect(list.totalPrice).toBe(3.59);
})

