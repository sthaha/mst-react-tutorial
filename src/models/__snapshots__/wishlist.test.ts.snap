// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add items to wishlist 1`] = `
Object {
  "items": Array [
    Object {
      "name": "foobar",
      "price": 1.56,
      "url": "",
    },
  ],
}
`;

exports[`can store patches of changes to wishlist 1`] = `
Object {
  "items": Array [
    Object {
      "name": "something",
      "price": 1.56,
      "url": "",
    },
  ],
}
`;

exports[`can store patches of changes to wishlist 2`] = `
Array [
  Object {
    "op": "add",
    "path": "/items/0",
    "value": Object {
      "name": "something",
      "price": 1.56,
      "url": "",
    },
  },
]
`;
